### PR TITLE
Fixes default distance thesholds and adds many test cases

### PR DESCRIPTION
**Pull request by apetkau**
* Sets the default parameter `--pd_distm hamming` instead of *scaled* as this better matches the default thresholds `--gm_thresholds 10,5,0` (those thresholds represent hamming distances between MLST profiles).
* Updates the regex for checking `--gm_thresholds` to be more specific.
* Adds many test cases for different combinations of `--gm_tresholds` and verifies cluster addresses are produced correctly/error checking works.
* Adds test cases for passing alleles as md5 hashes and counting/not counting missing data in distance counts.

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).

### CODE REVIEW

**Review by kylacochrane — COMMENTED** [review]
This looks great Aaron - so many tests 😸 

My only comment: Is it worth adding a consistency test case with mismatched pd_distm and gm_thresold? 

**E.g.** pd_distm = "scaled" and gm_thresold = "10,5,0"
**E.g.** pd_distm = "hamming" and gm_thresold = "0.8,0.15,0.5"

And maybe add a stdout for this error?

**Comment by apetkau** [comment]
@kylacochrane this is a really great suggestion. Thanks so much 😄. Implemented in 065b92b03faad8ea1e4984c307013a55c6860f8d

File: tests/data/samplesheets/samplesheet-hamming.csv [hunk 1]
```diff
@@ -0,0 +1,4 @@
+sample,mlst_alleles,metadata_1,metadata_2,metadata_3,metadata_4,metadata_5,metadata_6,metadata_7,metadata_8
+sample1,https://raw.githubusercontent.com/phac-nml/gasclustering/fix/distances/tests/data/reports/case-hamming/sample1.mlst.subtyping.json,,,,,,,,
+sample2,https://raw.githubusercontent.com/phac-nml/gasclustering/fix/distances/tests/data/reports/case-hamming/sample2.mlst.subtyping.json,,,,,,,,
+sample3,https://raw.githubusercontent.com/phac-nml/gasclustering/fix/distances/tests/data/reports/case-hamming/sample3.mlst.subtyping.json,,,,,,,,
```
Comment (emarinier) [inline tests/data/samplesheets/samplesheet-hamming.csv:4]: I think these file paths (`fix/distances`) may disappear when it's merged into dev and the branch deleted. You may want to either immediately change them after merging into dev, or preemptively push the test data to dev and update these to point there.

File: tests/pipelines/main.nf.test [hunk 2]
```diff
@@ -66,6 +69,139 @@ nextflow_pipeline {
...
+                gm_thresholds = "2,1,0"
+
+                metadata_1_header = "myheader_1"
+                metadata_2_header = "myheader_2"
+                metadata_3_header = "myheader_3"
+                metadata_4_header = "myheader_4"
+                metadata_5_header = "myheader_5"
+                metadata_6_header = "myheader_6"
+                metadata_7_header = "myheader_7"
+                metadata_8_header = "myheader_8"
```
Comment (emarinier) [inline tests/pipelines/main.nf.test:89]: I think you can leave this in if you want, but you could also have them as the defaults (and not explicitly test them for test compactness / readability). You'll have to update the ArborView HTML check later in this test though. Either way is fine.

File: tests/pipelines/main.nf.test [hunk 3]
```diff
@@ -66,6 +69,139 @@ nextflow_pipeline {
...
+
+            // Check that the ArborView output is created
+            def actual_arborview = path("$launchDir/results/ArborView/clustered_data_arborview.html")
+            assert actual_arborview.text.contains("id\\taddress\\tlevel_1\\tlevel_2\\tlevel_3\\tmyheader_1\\tmyheader_2\\tmyheader_3\\tmyheader_4\\tmyheader_5\\tmyheader_6\\tmyheader_7\\tmyheader_8\\nsample1\\t1.1.1\\t1\\t1\\t1\\t\\t\\t\\t\\t\\t\\t\\t\\nsample2\\t1.1.2\\t1\\t1\\t2\\t\\t\\t\\t\\t\\t\\t\\t\\nsample3\\t2.2.3\\t2\\t2\\t3\\t\\t\\t\\t\\t\\t\\t\\t\\n")
+
+            // compare IRIDA Next JSON output
+            def iridanext_json = path("$launchDir/results/iridanext.output.json").json
+            def iridanext_global = iridanext_json.files.global
+            def iridanext_samples = iridanext_json.files.samples
+            def iridanext_metadata = iridanext_json.metadata.samples
```
Comment (emarinier) [inline tests/pipelines/main.nf.test:128]: So, I know these lines have been propagated for a little while in a few tests, but they're not actually doing anything. It's just defining variables that aren't used.

File: tests/data/samplesheets/samplesheet-hash-missing.csv [hunk 4]
```diff
@@ -0,0 +1,4 @@
+sample,mlst_alleles,metadata_1,metadata_2,metadata_3,metadata_4,metadata_5,metadata_6,metadata_7,metadata_8
+sample1,https://raw.githubusercontent.com/phac-nml/gasclustering/fix/distances/tests/data/reports/case-hash-missing/sample1.mlst.subtyping.json,,,,,,,,
+sample2,https://raw.githubusercontent.com/phac-nml/gasclustering/fix/distances/tests/data/reports/case-hash-missing/sample2.mlst.subtyping.json,,,,,,,,
+sample3,https://raw.githubusercontent.com/phac-nml/gasclustering/fix/distances/tests/data/reports/case-hash-missing/sample3.mlst.subtyping.json,,,,,,,,
```
[the remark this answers is not in the document]
Comment (apetkau) [reply tests/data/samplesheets/samplesheet-hash-missing.csv:4]: Thanks. Will change after merge.

File: tests/pipelines/main.nf.test (hunk 2 above)
Comment (apetkau) [reply tests/pipelines/main.nf.test]: That's a good point. I've updated to check the default metadata headers instead in this test in 875113b41bd396fd15b1dbfac671830345ff6e32

File: tests/pipelines/main.nf.test (hunk 3 above)
Comment (apetkau) [reply tests/pipelines/main.nf.test:134]: Thanks. I completely forgot about adding any checks for data in the IRIDA Next JSON. This has been added here 5e2e49edd57b44ea9111b574cd4ff3bd5dd52407

**Review by emarinier — APPROVED** [review]

**Review by kylacochrane — APPROVED** [review]

**Pull request merged**